src/Mono.Android/Android.Runtime/AndroidEnvironment.cs

```diff
@@ -344,18 +344,60 @@ static IWebProxy GetDefaultProxy ()
 #if !MONOANDROID1_0
 		[DynamicDependency (DynamicallyAccessedMemberTypes.PublicParameterlessConstructor, typeof (Xamarin.Android.Net.AndroidMessageHandler))]
 #endif
-		static object? GetHttpMessageHandler ()
+		static object GetHttpMessageHandler ()
 		{
-			var envvar = Environment.GetEnvironmentVariable ("XA_HTTP_CLIENT_HANDLER_TYPE")?.Trim ();
+			var handlerTypeName = Environment.GetEnvironmentVariable ("XA_HTTP_CLIENT_HANDLER_TYPE")?.Trim ();
 			Type? handlerType = null;
-			if (!String.IsNullOrEmpty (envvar))
-				handlerType = Type.GetType (envvar, false);
-			else
-				return null;
-			if (handlerType == null)
-				return null;
+			if (!String.IsNullOrEmpty (handlerTypeName))
+				handlerType = Type.GetType (handlerTypeName, throwOnError: false);
+
 			// We don't do any type checking or casting here to avoid dependency on System.Net.Http in Mono.Android.dll
-			return Activator.CreateInstance (handlerType);
+			if (handlerType is null || !IsAcceptableHttpMessageHandlerType (handlerType)) {
+				handlerType = GetFallbackHttpMessageHandlerType ();
+			}
+
+			return Activator.CreateInstance (handlerType)
+				?? throw new InvalidOperationException ($"Could not create an instance of HTTP message handler type {handlerType.AssemblyQualifiedName}");
+		}
+
+		static bool IsAcceptableHttpMessageHandlerType (Type handlerType)
+		{
+#if !MONOANDROID1_0
+			if (Extends (handlerType, "System.Net.Http.HttpClientHandler")) {
+				// It's not possible to construct HttpClientHandler in this method because it would cause infinite recursion
+				// as HttpClientHandler's constructor calls the GetHttpMessageHandler function
+				Logger.Log (LogLevel.Warn, "MonoAndroid", $"The type {handlerType.AssemblyQualifiedName} cannot be used as the native HTTP handler because it is derived from System.Net.Htt.HttpClientHandler. Use a type that extends System.Net.Http.HttpMessageHandler instead.");
+				return false;
+			}
+#endif
+			if (!Extends (handlerType, "System.Net.Http.HttpMessageHandler")) {
+				Logger.Log (LogLevel.Warn, "MonoAndroid", $"The type {handlerType.AssemblyQualifiedName} set as the default HTTP handler is invalid. Use a type that extends System.Net.Http.HttpMessageHandler.");
+				return false;
+			}
+
+			return true;
+		}
+
+		static bool Extends (Type handlerType, string typeName, string assemblyName = "System.Net.Http")
+		{
+			while (handlerType != null) {
```

Comment:
This method body seems odd; why not:

```csharp
static bool Extends (Type handlerType, string assemblyQualifiedTypeName)
{
    var baseType = Type.GetType (assemblyQualifiedTypeName, throwOnError:false);
    if (baseType == null)
        return false;
    return baseType.IsAssignableFrom (handlerType);
}
```

…but probably move the `Type.GetType()` into the caller scope…?

My *guess* as to using the `while` loop is to avoid the `Type.GetType()` invocation; is it *that* expensive?  It's not like the `while` loop *prevents* the equivalent of `typeof(System.Net.Http.HttpMessageHandler)` from running, as `handlerType.BaseType` will *require* that it be loaded if it's in the inheritance chain.

If e.g. `System.Net.Http.HttpMessageHandler ` is linked away, then `handlerType` cannot possibly extend it.

Reply:
You're right, that's much cleaner. Thanks for the suggestion.